src/Control/Effect/Interpret.hs

```diff
@@ -0,0 +1,82 @@
+{-# LANGUAGE GeneralizedNewtypeDeriving, MultiParamTypeClasses, RankNTypes, TypeOperators, UndecidableInstances #-}
+module Control.Effect.Interpret
+( runInterpret
+, InterpretC(..)
+, runInterpretState
+, InterpretStateC(..)
+) where
+
+import Control.Applicative (Alternative(..))
+import Control.Effect.Carrier
+import Control.Effect.Reader
+import Control.Effect.State
+import Control.Effect.Sum
+import Control.Monad (MonadPlus(..))
+import Control.Monad.Fail
+import Control.Monad.IO.Class
+import Control.Monad.Trans.Class
+
+-- | Interpret an effect using a higher-order function.
+--
+--   This involves a great deal less boilerplate than defining a custom 'Carrier' instance, at the expense of somewhat less performance. It’s a reasonable starting point for new interpretations, and if more performance or flexibility is required, it’s straightforward to “graduate” by replacing the relevant 'runInterpret' handlers with specialized 'Carrier' instances for the effects.
+--
+--   At time of writing, a simple passthrough use of 'runInterpret' to handle a 'State' effect is about five times slower than using 'StateC' directly.
```

Comment:
I fixed this with a [rewrite rule](https://github.com/isovector/polysemy/blob/546cf16e1b2146326e8e2908847c75c5c82b41e6/src/Polysemy/State.hs#L37-L40)

Reply:
That’s for `reinterpret`, which is quite a different case. Further, I think this is behaves-as-intended; if you want better performance, you should define a new `Carrier`.

Reply:
I’m of two minds about including this rewrite rule. On the one hand, I don’t want to encourage people to always use `interpret`; on the other hand, I am uncomfortable knowingly passing up a serious performance optimization. I think I would prefer to include the rule but include caveats that performance-critical effects should not use `interpret`.

Reply:
Am I correct in that we’d need to define a `REWRITE` rule per-handler for each effect that needed to fuse with Interpret?

Reply:
Yes. And since we already ship with good carriers for each of our provided effects, it’s not clear to me that anyone would get any benefit from this; i.e. you shouldn’t use `runInterpret` to pass-through `State` effects to an underlying `StateC`, you should just use the underlying `StateC` directly.